lib/timers.js

```diff
@@ -10,36 +10,92 @@ const kOnTimeout = Timer.kOnTimeout | 0;
+// map, or creating a new one, but it is trivial in comparison.
+// Removal from an object-property linked list is also constant-time as can be
+// seen in the lib/internal/linkedlist.js implementation.
+// Timeouts only need to process any timers due to currently timeout, which will
+// always be at the  beginning of the list for reasons stated above. Any timers
+// after the first one encountered that does not yet need to timeout will also
+// always be due to timeout at a later time.
 //
-// Because often many sockets will have the same idle timeout we will not
-// use one timeout watcher per item. It is too much overhead.  Instead
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
+// The technique is described (in lesser detail) within the libev manual:
 // http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#Be_smart_about_timeouts
 
-// Object containing all lists, timers
-// key = time in milliseconds
-// value = list
-var lists = {};
 
+// Object maps containing linked lists of timers, keyed and sorted by their
+// duration in milliseconds.
+// - key = time in milliseconds
+// - value = linked list
+const refedLists = {};
+const unrefedLists = {};
+// The difference between these two objects is that the former contains timers
+// that will keep the process open if they are the only thing left, while the
+// latter will not.
 
-// call this whenever the item is active (not idle)
-// it will reset its timeout.
-// the main function - creates lists on demand and the watchers associated
-// with them.
+
+// Schedule or re-schedule a timer.
+// Call this whenever the item is active (not idle).
+// The item must have been enroll()'d first.
 exports.active = function(item) {
+  insert(item, false);
+};
+
+// Internal APIs that need timeouts should use `_unrefActive()` instead of
+// `active()` so that they do not unnecessarily keep the process open.
+exports._unrefActive = function(item) {
+  insert(item, true);
+};
+
+
+// The underlying logic for scheduling or re-scheduling a timer.
+//
+// Appends a timer onto the end of an existing timers list, or creates a new
+// TimerWrap backed list if one does not already exist for the specified timeout
+// duration.
+function insert(item, unrefed) {
   const msecs = item._idleTimeout;
   if (msecs < 0 || msecs === undefined) return;
 
   item._idleStart = Timer.now();
 
-  var list;
+  const lists = unrefed ? unrefedLists : refedLists;
 
-  if (lists[msecs]) {
-    list = lists[msecs];
-  } else {
+  // Use an existing list if there is one, otherwise we need to make a new one.
+  var list = lists[msecs];
+  if (!list) {
+    // Make a new linked list of timers.
+    // The list object is a TimerWrap, a C++ handle backing the timers for
+    // efficiency, and the linked list is appended onto it.
     list = new Timer();
+    if (unrefed) list.unref();
```

Comment:
Prefer strict equality here so the full-codegen can emit an immediate boolean check instead of a ToBoolean IC.